packages/react/src/JsonFormsContext.tsx

```diff
@@ -144,6 +189,89 @@ export const JsonFormsStateProvider = ({
       })
     )
   );
+
+  useEffect(() => {
+    coreDispatch(
+      Actions.updateCore(data, schema, uischema, {
+        ajv,
+        validationMode,
+        additionalErrors,
+      })
+    );
+  }, [data, schema, uischema, ajv, validationMode, additionalErrors]);
+
+  /**
+   * A common pattern for users of JSON Forms is to feed back the data which is emitted by
+   * JSON Forms to JSON Forms ('controlled style').
+   *
+   * Every time this happens, we dispatch the 'updateCore' action which will be a no-op when
+   * the data handed over is the one which was just recently emitted. This allows us to skip
+   * rerendering for all normal cases of use.
+   *
+   * However there can be extreme use cases, for example when using Chrome Auto-fill for forms,
+   * which can cause JSON Forms to emit multiple change events before the parent component is
+   * rerendered. Therefore not the very recent data, but the previous data is fed back to
+   * JSON Forms first. JSON Forms recognizes that this is not the very recent data and will
+   * validate, rerender and emit a change event again. This can then lead to data loss or even
+   * an endless rerender loop, depending on the emitted events chain.
+   *
+   * To handle these edge cases in which many change events are sent in an extremely short amount
+   * of time we debounce them over a short amount of time. 10ms was chosen as this worked well
+   * even on low-end mobile device settings in the Chrome simulator.
+   */
+  const debouncedEmit = useCallback(
+    debounce((...args: any[]) => onChangeRef.current?.(...args), 10),
+    []
+  );
+
+  return (
+    <JsonFormsDataProvider
+      initState={initState}
+      onChangeHandler={debouncedEmit}
+      core={core}
+      coreDispatch={coreDispatch}
+    >
+      {children}
+    </JsonFormsDataProvider>
+  );
+};
+
+const JsonFormsControlledStateProvider = ({
+  children,
+  initState,
+  onChangeRef,
+}: JsonFormsStateProviderProps & { onChangeRef: any }) => {
+  const { data, schema, uischema, ajv, validationMode, additionalErrors } =
+    initState.core;
+
+  const onChangeHandler = useCallback(
+    (...args: any[]) => onChangeRef.current?.(...args),
+    []
+  );
```

Comment:
Is there a benefit of introducing this additional `useCallback`? Later where we use `onChangeHandler` we could just use `onChangeRef.current?` directly.